### PR TITLE
bpo-18533: Avoid RuntimeError from repr() of recursive dictview

### PR DESCRIPTION
**Pull request by bennorth**
(Created as per [recent comment](https://bugs.python.org/issue18533#msg308086) on bpo-18533.  See issue for further background and discussion.)

`dictview_repr()`: Use a `Py_ReprEnter()` / `Py_ReprLeave()` pair to check for recursion, and produce `"..."` if so.

`test_recursive_repr()`: Check for the correct string (containing `"..."`) rather than a `RuntimeError`.

`test_deeply_nested_repr()`: Add new test, replacing the original `test_recursive_repr()`.  It checks that a `RuntimeError` is raised in the case of a non-recursive but deeply nested structure.  (Very similar to what `test_repr_deep()` in `test/test_dict.py` does for a normal dict.)

https://bugs.python.org/issue18533

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

**Comment by vstinner** [comment]
Hum, this bug has already been fixed in https://bugs.python.org/issue32137 by the commit 1fb72d2ad243c965d4432b4e93884064001a2607, no?

cc @serhiy-storchaka

**Comment by bennorth** [comment]
The fix for bpo-32137 changed a segfault into a `RecursionError`.  This PR proposes changing that into a friendlier `"..."`-style representation, i.e., no exception is raised.

(I've just noticed that my commit message incorrectly refers to `RuntimeError`, sorry, as this is what it was last time I looked at this.  I can force-push a commit with better message if this PR is approved in principle.)

File: Objects/dictobject.c [hunk 1]
```diff
@@ -3781,14 +3781,22 @@ static PyObject *
 dictview_repr(_PyDictViewObject *dv)
 {
     PyObject *seq;
-    PyObject *result;
+    PyObject *result = NULL;
+    Py_ssize_t rc;
+
+    rc = Py_ReprEnter((PyObject *)dv);
+    if (rc != 0)
```
Comment (serhiy-storchaka) [inline Objects/dictobject.c]: Add braces.

File: Objects/dictobject.c [hunk 2]
```diff
@@ -3781,14 +3781,22 @@ static PyObject *
...
+        return rc > 0 ? PyUnicode_FromString("...") : NULL;
 
     seq = PySequence_List((PyObject *)dv);
     if (seq == NULL)
```
Comment (serhiy-storchaka) [inline Objects/dictobject.c]: Add braces.

File: Lib/test/test_dictviews.py [hunk 3]
```diff
@@ -202,6 +203,18 @@ def test_items_set_operations(self):
     def test_recursive_repr(self):
         d = {}
         d[42] = d.values()
+        r = repr(d)
+        self.assertIsInstance(r, str)
```
Comment (serhiy-storchaka) [inline Lib/test/test_dictviews.py]: Is this needed?

**Review by serhiy-storchaka — COMMENTED** [review]
LGTM. Just fix a style and add a news entry.

What about OrderedDict? Is it fixed by this change?

**Comment by bennorth** [comment]
@serhiy-storchaka --- I've implemented those style requests, and added a news entry.

Without this PR, `OrderedDict` raises a `RecursionError`:
```
>>> from collections import OrderedDict
>>> d = OrderedDict()
>>> d[42] = d.values()
>>> repr(d)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RecursionError: maximum recursion depth exceeded while getting the repr of an object
```
whereas with this PR:
```
>>> repr(d)
'OrderedDict([(42, odict_values([...]))])'
```
The PR now adds a test for the new behaviour of `OrderedDict`.

File: Objects/dictobject.c [hunk 4]
```diff
@@ -3781,14 +3781,24 @@ static PyObject *
...
 {
     PyObject *seq;
-    PyObject *result;
+    PyObject *result = NULL;
+    Py_ssize_t rc;
+
+    rc = Py_ReprEnter((PyObject *)dv);
+    if (rc != 0) {
+        return rc > 0 ? PyUnicode_FromString("...") : NULL;
+    }
```
Comment (serhiy-storchaka) [inline Objects/dictobject.c]: I think empty lines are not needed after `}`. `}` itself adds enough vertical space.

File: Lib/test/test_collections.py [hunk 5]
```diff
@@ -1916,6 +1916,22 @@ def test_helper_function(self):
         self.assertEqual(dict(c), {'a': 5, 'b': 2, 'c': 1, 'd': 1, 'r':2 })
 
 
+################################################################################
+### OrderedDict
+################################################################################
+
+class TestOrderedDict(unittest.TestCase):
+
+    def test_recursive_repr(self):
```
Comment (serhiy-storchaka) [inline Lib/test/test_collections.py]: Move this test into test_ordered_dict.py.

File: Lib/test/test_collections.py [hunk 6]
```diff
@@ -1916,6 +1916,22 @@ def test_helper_function(self):
...
+        d = OrderedDict()
+        d[42] = d.values()
+        r = repr(d)
+        self.assertEqual(r, 'OrderedDict([(42, odict_values([...]))])')
```
Comment (serhiy-storchaka) [inline Lib/test/test_collections.py]: The concrete implementation is implementation dependent. For example in PyPy3 it is `OrderedDict([(42, _OrderedDictValuesView(...))])`. I think it enough to test that repr() doesn't raise an exception in other implementations.
```python
r = repr(d)
if test.support.check_impl_detail(cpython=True):
    self.assertEqual(r, 'OrderedDict([(42, odict_values([...]))])')
```

**Comment by bennorth** [comment]
Thanks for feedback.  New version pushed addressing comments:
* Remove vertical whitespace;
* Avoid relying on particular implementation in tests;
* Move `OrderedDict` test to more sensible location.

Also word 'news' entry more concisely.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by bennorth** [comment]
Is there anything further I should be doing to help this along?  The [failing AppVeyor build](https://ci.appveyor.com/project/python/cpython/build/3.7.0a0.9452) looks unrelated as it's a failure in `distutils.tests.test_bdist_wininst.BuildWinInstTestCase`.

**Comment by bennorth** [comment]
@terryjreedy — Thanks for bringing this up-to-date with `master`; good that the AppVeyor build is now OK.

**Comment by orsenthil** [comment]
I merged it, @bennorth. All the requirements were met and thanks for your contribution.

**Comment by bennorth** [comment]
@orsenthil — thanks for merging.

As suggested in [comment on issue](https://bugs.python.org/issue18533#msg310785), I've created #5357 to backport this to 2.7.

**Pull request merged**